server/src/main/java/org/elasticsearch/search/aggregations/metrics/PercentilesMethod.java

```diff
@@ -65,4 +86,165 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
         return parseField.getPreferredName();
     }
+
+    /**
+     * A small config object that carries algo-specific settings.  This allows the factory to have
+     * a single unified constructor for both algos, but internally switch execution
+     * depending on which algo is selected
+     */
+    public abstract static class Config implements ToXContent, Writeable {
+        private final PercentilesMethod method;
+
+        Config(PercentilesMethod method) {
+            this.method = method;
+        }
+
+        public static Config fromStream(StreamInput in) throws IOException {
+            PercentilesMethod method = PercentilesMethod.readFromStream(in);
+            if (method.equals(PercentilesMethod.TDIGEST)) {
```

Comment:
I really don't like this if(method.equals(TDIGEST) {...} else if (method.equals(HDR) {...} else { throw } pattern.  We use it in a lot of places, and if we ever add another percentile method, we'll have to find and touch each of them.  Plus, the throw is always unreachable now, since we'd have already thrown trying to create a bogus enum instance.

In this case, it's pretty trivial to put a method on the enum that takes a `StreamInput` and returns a `Config`, and I think we should do that.  We should at least think about applying this logic to the other places we have this type of construct.

Reply:
Pushed a change that takes care of the factory/builder instances of these as discussed on slack.  I'm not sure _this_ particular one is easily solved though.  Note that this is a static method on the abstract Config object, one level "inside" the `PercentilesMethod` enum

The higher-level `PercentilesMethod` has a `readFromStream()` which is in use today to serialize the enum (as well as both `compression` and `numSigFig` in the builder).  The builder already has to incorporate version checks for wire BWC, so I was hesitant to lump the `Config` serialization into the existing `PercentilesMethod` because it would also need version checks.

(Also, conceptually we're in a slightly odd place where the "new way" is to serialize the Config object which holds a `PercentileMethod` enum, so asking the enum to deserialize a Config seems incorrect.  But that's more of an artifact of the current hierarchy which should be cleaner once things are removed post-deprecation)

Otherwise, I'm happy to move this up a level... but I think we'll still need these if/else at this point in the code.  Since this isn't a NamedWriteable style situation, we have to inspect the serialized method to determine what kind of object needs to be built next.  I don't see any way around that?